### ISSUE TITLE
Cookie Consent front-end not accessible to screen readers

### ISSUE THREAD

**Issue by alexstine** [Automattic/jetpack#30230]
### Impacted plugin

Jetpack

### Quick summary

The Cookie Consent banner on the front-end is not accessible to screen readers. It is loaded at the end of the DOM making it undiscoverable by blind users.

### Steps to reproduce

1. Enable a screen reader such as NVDA on Windows or Voiceover on Mac.
2. Enable the Jetpack Cookie Consent banner.
3. View the front-end.
4. Notice how the banner is loaded at the end of the DOM where a user would have to navigate through the entire page before locating it.

### A clear and concise description of what you expected to happen.

I would expect the Cookie Consent banner to be marked up as a real dialog.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog

Focus should be managed when the dialog launches and when closing the dialog. This is the only way to ensure screen readers get the same information that sighted people receive.

### What actually happened

See above.

### Impact

All

### Available workarounds?

No and the platform is unusable

**Comment by cuemarie** [comment]
📌 SCRUBBING : RESULT - *Replicated / Could Not Replicate / Uncertain*
- Tested on Simple - **Replicated**
- Tested on Atomic – **Replicated**

📌 FINDINGS/SCREENSHOTS/VIDEO
-  On both AT and Simple sites, using either the Cookies & Consent widget (at top or bottom of site) as well as the default banner added to free sites displaying ads, I can replicate the issue as described.

#### Screenshots/Recordings

![Markup on 2023-04-20 at 16:45:22](https://user-images.githubusercontent.com/27249804/233509310-8805ed32-e7a5-4d25-bf84-6a53bed4fb68.png)

Can also see in Accessibility Inspector that the Widget is the very last landmark on the page:

![Markup on 2023-04-20 at 16:49:24](https://user-images.githubusercontent.com/27249804/233509584-ae4e69ac-73b2-4ebc-82c2-12dabf020bb4.png)

**Free Site displaying Ads example**

![Markup on 2023-04-20 at 16:52:34](https://user-images.githubusercontent.com/27249804/233509835-a96eeeda-cb02-471f-b4e3-6620ec5f3074.png)

📌 ACTIONS
- Assigned to Other

**Comment by jeherve** [comment]
cc @alshakero, who worked on the block. Do you think you could take a look?

Thank you!

**Comment by alshakero** [comment]
Hey everyone! This widget is deprecated. [We have a cookie consent **block** now](https://wordpress.com/support/cookie-consent-block/). I did a cursory accessibility audit for that one, and it looks accessible.

**Comment by alexstine** [comment]
CC: @ryelle. See above.

**Comment by cuemarie** [comment]
Hey @alshakero and others, thanks for directing to the Cookie/Consent block!  A couple of questions:

First, the original report states:

> Focus should be managed when the dialog launches and when closing the dialog. This is the only way to ensure screen readers get the same information that sighted people receive.

I do see the consent block now appears as a `dialog` in Accessibility properties, so that seems to address this issue.

Second, the report also states:

> Notice how the banner is loaded at the end of the DOM where a user would have to navigate through the entire page before locating it.

The consent dialog does still appear at the end of the page, rather than at the beginning, when I inspect the accessibility properties. Are there any concerns/considerations to be had here, that would prioritize getting cookie consent from screen reader users earlier in their experience of the site?

![Markup on 2023-05-02 at 14:44:27](https://user-images.githubusercontent.com/27249804/235792808-208874f9-f0c8-443a-a8fe-f131ed922115.png)

Lastly, the Cookie/Consent block is only available for use on sites using the Site Editor. While AT sites on classic themes can install a plugin, I believe simple sites on classic themes can only use the deprecated widget.  I'm not sure if that warrants any other prioritization of the widget issue, but wanted to mention it.

Thanks all!

**Comment by alexstine** [comment]
The dialog can appear at the end of the DOM as long as focus is managed properly.

**Comment by LukaszJaro** [comment]
I ran into this issue with a alert banner that was right below the header, so it was not the first element in the DOM tree. We couldn't change the design of the component at the moment and move it to the very top so I had to use some JS along with with modifying the role from alert to alertdialog, labelledby, aria-modal, basically the markup from w3 org example https://www.w3.org/WAI/ARIA/apg/patterns/alertdialog/examples/alertdialog/

I had to set focus event listener on the very first element in the DOM which happened to be the skip links, as soon as user tabbed to first dom element, it would instantly skip to the alertdialog and focus on the close button, from that point the focus would be trapped until user closed it, or pressed escape key. After alert dialog was closed, the focus would go back to the skip to content link.

I had alot of trouble finding accessible examples of a banner that was not at the very top and also wondered why we couldn't leave it as a alert role..all I found was the below exmaples but they have the cookie banner html at the very top and one of them uses css to make the banner fixed to bottom.

- https://codepen.io/bellangerq/full/vYOwpvm
- https://techservicesillinois.github.io/accessibility/examples/notice/index.html

**Comment by alexstine** [comment]
@LukaszJaro Even in your fix, it wouldn't have worked for most screen reader users. `onFocus` events do not fire in screen reader browse/virtual mode where users use arrow keys to navigate. Events only pass to JavaScript in two ways.
1. Tab/Shift+Tab and a couple other keys.
2. If the screen reader is set to forms/edit mode sending all keyboard events to the browser, outside of screen reader virtual document mode.

So, if I came to your website and only used my arrow keys to navigate, focus would have never moved.

Using alert roles are annoying and should never be over used. Only for truly important information.

The cookie notice doesn't have to change DOM position, developers simply need to manage focus.

`<div role="dialog" tabindex="-1" aria-labelledby="dialog-title" aria-describedby="dialog-desc"></div>

When the dialog renders, move focus to it. You can do this with `React.useRef` and `React.useEffect` for example.

**Comment by LukaszJaro** [comment]
Thank you @alexstine for the valuable feedback. I did not do the full accessibility testing to that extent since we have a accessibility expert on our team that performs the tests and either passes or fails. Also the solution we have now is temporary and it will actually be converted to a real modal.
